### DISCUSSION TITLE
Getting the Activated Workflow Instance

### DISCUSSION THREAD

**Discussion by comicon1234** [elsa-workflows/elsa-core#487] [Q&A]
Is there a way to see when a workflow has activated and get the workflow instance ID to use that somewhere else. I am trying to use elsa to do the heavy lifting of a Request approval workflow and I have a ton of data from the request in another table. I would like to create a relationship between the Elsa workflow table and the Request data table when a new request is submitted.

**Comment by comicon1234** [comment]
Would creating a custom activity be the way to go? I see that I can get the Workflow Execution context when creating a custom activity.

**Comment by sfmskywalker** [comment]
Not knowing the full details of your requirements, I would say that in general, the best way to go about this would be something along the following lines:

* When a new request is submitted, you fire some event (perhaps using Mediatr), unless you are already implementing some form of even handler for when a new request is submitted.
* Regardless of the above, use Elsa's Workflow Trigger API to either start new or resume existing workflows that are (to be) correlated with the request (usually by ID, but this is up to you).
* The workflow does its thing. It can interact with the request and implement logic based on the request's properties.

What I would do is indeed create a custom activity, e.g. `RequestSubmitted`, which is an activity that acts as a trigger (causing the workflow to be suspended unless it's the first activity in the workfow to execute).

**Elsa 1**
From your application code, as soon as you identified that a new request is submitted, you do these things:
1. Get all workflow definitions from `IWorkflowRegistry` that start with the custom `RequestSubmitted` activity.
1. Start each workflow using `IWorkflowInvoker.StartAsync`.
1. Get all workflow instances from `IWorkflowInstanceStore` that are blocked on the `RequestSubmitted` activity.
1. Resume each workflow using `IWorkflowInvoker.ResumeAsync`.

Look at `TriggerRequestHandler` in `Elsa.Activities.Http.RequestHandlers.Handlers` for an example of how workflows are started and resumed based on incoming HTTP requests.

**Elsa 2**
The following is subject to change, but right now you would do this:
1. Implement a `Trigger` and its `TriggerProvider` as a way to index workflows so that you can select them when it is time to trigger them (when a request is submitted). The trigger could be named e.g. `RequestSubmittedTrigger` and have a `RequestId` field.
1. Implement a `RequestSubmittedTriggerProvider` by inheriting from `TriggerProvider<RequestSubmittedTrigger, RequestSubmitted>`.
1. When a request is submitted, use the `IWorkflowRunner` service to `TriggerWorkflowsAsync<RequestSubmittedTrigger>(x => x.RequestId == null || x.RequestId == request.Id)`, which will either start a new workflow of none exists yet for the submited request, or resume existing ones that are blocked on the `RequestSubmitted` activity.

The above is the gist of it, and there are some details left out because the trigger API in Elsa 2 is still unstable, so details will change. Once everything is fleshed out there, all of this will be documented. Until that time, I would encourage you to check out the sample applications and existing activity implementations. And when you have specific questions, feel free to drop them here or by creating new issues or by asking questions on SO.

**Reply by AmineMraihi** [reply]
from elsa 1
step 3:
I'm currently facing issues, 
to clarify a bit more, I m trying to implement Elsa into my project, and I want to use an implementation of IWorkflowInstanceStore but coulding
at the end I would like the possibility to do this: await _workflowInstanceStore.FindManyAsync etc
can you help?
thanks in advance.

**Reply by sfmskywalker** [reply]
Can you clarify what you mean when you say that you want to use an implementation of `IWorkflowInstanceStore` but that you could not? Why weren't you able to use it? Did you get an error, or are you saying you don't know how to use an implementation of it? If so, what implementation are you trying to use? What did you try specifically and how did it not work out?

The more info you give, the higher the chance I can help 😄

**Reply by AmineMraihi** [reply]
Hello, thanks for coming back to me,
I actually found a solution for my problem.
I created a project where I wanted to use Elsa methods such as retreiving workflow instances and to do so I have to have an instance of IWorkflowInstanceStore and I successfully did it.
It may not be the best way to do it in my opinion but happy if you have other ideas.
I created a helper class in which i have a property of type IWorkflowInstanceStore
and in my startup.cs I set it up using: serviceProvider.GetRequiredService<IWorkflowInstanceStore>();
and then whenever i want to use that instance I just call my helper class (it is static)
is that the best way?  
Thanks in advance,
Amine

**Reply by sfmskywalker** [reply]
I see. It might cause issues down the line, since `IWorkflowInstanceStore` is a scoped service. This means that when the service provider goes out of scope (e.g. at the end of an HTTP request), all services and their dependencies within the same scope will be disposed.

The usual way to do things is as follows.

Let's say you have two projects:

1. Web project
2. Library project

And the Library project is a dependency of the Web project, 
Now, if your Library project has a class called e.g. `MyLibraryService`, then it can work with `IWorkflowInstanceStore` like this:

```csharp
public class MyLibraryService : IMyLibraryService
{
   private readonly IWorkflowInstanceStore _workflowInstanceStore;

   public MyLibraryService(IWorkflowInstanceStore workflowInstanceStore)
   {
      _workflowInstanceStore = workflowInstanceStore;
   }
}
```

For this to work, make sure to register `MyLibraryService` as a scoped service in the `Startup` class of the Web project:

```csharp
services.AddScoped<IMyLibraryService, MyLibraryService>();
```

This way, your library service has the same lifetime as `IWorkflowInstanceStore` and you can inject in other scoped services, including controllers.

**Reply by AmineMraihi** [reply]
Hello, 
sorry for the late reply.
Yes indeed it works (didn't know why hadn't thought about this for the time :( ).
Again I would like to thank you for your kind support and help.
Amine.

**Reply by sfmskywalker** [reply]
No worries 😄 
Thanks for confirming that works.

**Comment by comicon1234** [comment]
Thanks, you thoroughly answered my question.
